# Exploring Ebay Car Sales Data

In this project, I work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. The dataset was originally scraped and uploaded to Kaggle by user orgesleka.

I've made a few modifications from the original dataset: I sampled 50,000 data points from the full dataset.

In [1]:
import pandas as pd

autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [2]:
# Show dataset head
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [3]:
# Show dataset information
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

From this output we got:
1. Total number of column is 20
2. Some columns contained null value (e.g: vehicleType, gearbox, model, fuelType, notRepairedDamage)

### 1. Cleaning Column Names

From the work we did in the last screen, we can make the following observations:

- Some columns have null values, but none have more than ~20% null values.
- The column names use camelcase instead of Python's preferred snakecase, which means it can't just replace spaces with underscores.

Next step is change the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [4]:
# Renaming existing columns name to snakecase

change_col = {
    "dateCrawled": "date_crawled",
    "offerType": "offer_type",
    "vehicleType": "vehicle_type",
    "yearOfRegistration": "registration_year",
    "powerPS": "power_PS",
    "monthOfRegistration": "registration_month",
    "fuelType": "fuel_type",
    "notRepairedDamage": "not_repaired_damage",
    "dateCreated": "ad_created",
    "nrOfPictures": "nr_of_pictures",
    "postalCode": "postal_code",
    "lastSeen": "last_seen",
}

autos.rename(columns=change_col, inplace=True)

In [5]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand not_repaired_damage           ad_created  nr_of_pictures  \
0     peugeot                nein  2016-03-26 00:00:00               0   
1         bmw                nein  2016-04-04 00:00:00               0   
2  volkswagen                nein  2016-03-26 00:00:00               0   
3       smart                nein  2016-03-12 00:00:00               0   
4        ford                nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

From dataset above, I've made clean the columns name to snake_case to suit python's convention.

### 2. Initial Exploration and Cleaning.

I will do some basic data exploration to determine what other cleaning tasks need to be done. Initially I will look for:

- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
- Examples of numeric data stored as text which can be cleaned and converted.

In [6]:
# Look for descriptive statistics for all columns

autos.describe()

registration_year      power_PS  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

From table above, it is clearly that:
1. Column **"nr_of_pictures"** needs to be dropped because it is only containing zero value.
2. Column **"year_of_registration"**, **"power_PS"**, **"postal_code"** need more investigation since the max value is very high.

In [7]:
# Numeric data stored as text that needs to be cleaned.
autos[["price", "odometer"]].head()

price   odometer
0  $5,000  150,000km
1  $8,500  150,000km
2  $8,990   70,000km
3  $4,350   70,000km
4  $1,350  150,000km

It is noted that for column "price" and "odometer" is stored as a text, therefore it needs to convert to numeric.

In [8]:
# Remove any non-numeric characters
autos["price"] = autos["price"].str.replace(",", "")
autos["price"] = autos["price"].str.replace("$", "")
autos["odometer"] = autos["odometer"].str.replace("km", "")
autos["odometer"] = autos["odometer"].str.replace(",", "")

<ipython-input-8-6040a536a7bd>:3: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  autos["price"] = autos["price"].str.replace("$", "")


In [9]:
# Convert the column to a numeric dtype.
autos[["price", "odometer"]] = autos[["price", "odometer"]].astype(int)

autos[["price", "odometer"]].head(3)

price  odometer
0   5000    150000
1   8500    150000
2   8990     70000

"**odometer**" column is not clearly described, hence it is needed to change to "**odometer_km**"

In [10]:
# Change "odometer" column
autos.rename(columns={"odometer": "odometer_km"}, inplace=True)

In [11]:
autos[["price", "odometer_km"]].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   price        50000 non-null  int32
 1   odometer_km  50000 non-null  int32
dtypes: int32(2)
memory usage: 390.8 KB


### 3. Exploring the Odometer and Price Columns

From the last screen, it is known that there are a number of text columns where almost all of the values are the same (seller and offer_type). Also it needs to convert the price and odometer columns to numeric types and renamed odometer to odometer_km.

I will continue exploring the data, specifically looking for data that doesn't look right. Start by analyzing the odometer_km and price columns

In [12]:
# Exploring Odometer Column
print(autos["odometer_km"].unique().shape)
print("\n")
print(autos["odometer_km"].describe())
print("\n")
print(autos["odometer_km"].value_counts().sort_index(ascending=False))

(13,)


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64


In [13]:
# Exploring Price Column
print(autos["price"].unique().shape)
print("\n")
print(autos["price"].describe())
print("\n")
print(autos["price"].value_counts().sort_index(ascending=False))

(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
            ... 
5              2
3              1
2              3
1            156
0           1421
Name: price, Length: 2357, dtype: int64


In the Price Column is a bit odd, it shown the car price is ranging from 0 to 99,999,999. Therefore, it is needed to remove the outliers, so I range the price from 1,000 to 30,000,000.

In [14]:
# Range Price from 1,000 to 30,000,000
autos = autos[autos["price"].between(1000, 30000000)]
autos.describe()

price  registration_year      power_PS    odometer_km  \
count  3.863900e+04       38639.000000  38639.000000   38639.000000   
mean   9.996434e+03        2005.676596    128.927690  122774.010715   
std    2.018225e+05          86.671288    215.888529   40801.754668   
min    1.000000e+03        1000.000000      0.000000    5000.000000   
25%    2.200000e+03        2001.000000     80.000000  100000.000000   
50%    4.350000e+03        2005.000000    116.000000  150000.000000   
75%    8.950000e+03        2009.000000    160.000000  150000.000000   
max    2.732222e+07        9999.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        38639.000000         38639.0  38639.000000  
mean             5.998137             0.0  51690.971790  
std              3.591360             0.0  25691.970827  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  31224.000000  
50%              6.000000             0.0  50858.000000  
75%              9.000000             0.0  72537.000000  
max             12.000000             0.0  99998.000000

### 4. Exploring the date columns

Now move on to the date columns and understand the date range the data covers.

In [15]:
print(autos.info())
autos.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 38639 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   date_crawled         38639 non-null  object
 1   name                 38639 non-null  object
 2   seller               38639 non-null  object
 3   offer_type           38639 non-null  object
 4   price                38639 non-null  int32 
 5   abtest               38639 non-null  object
 6   vehicle_type         35880 non-null  object
 7   registration_year    38639 non-null  int64 
 8   gearbox              37201 non-null  object
 9   power_PS             38639 non-null  int64 
 10  model                37016 non-null  object
 11  odometer_km          38639 non-null  int32 
 12  registration_month   38639 non-null  int64 
 13  fuel_type            36276 non-null  object
 14  brand                38639 non-null  object
 15  not_repaired_damage  32863 non-null  object
 16  ad_c

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_PS   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand not_repaired_damage           ad_created  nr_of_pictures  \
0     peugeot                nein  2016-03-26 00:00:00               0   
1         bmw                nein  2016-04-04 00:00:00               0   
2  volkswagen                nein  2016-03-26 00:00:00               0   
3       smart                nein  2016-03-12 00:00:00               0   
4        ford                nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Now move on to the date columns and understand the date range the data covers.

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. 

Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, it needs to convert the data into a numerical representation so it can be to understand quantitatively.

#### Get a percentage for date_crawled column

In [16]:
date_crawled = (
    autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False) * 100
).sort_index()

print(date_crawled)

2016-03-05    2.554414
2016-03-06    1.387199
2016-03-07    3.511996
2016-03-08    3.266130
2016-03-09    3.248014
2016-03-10    3.330832
2016-03-11    3.279070
2016-03-12    3.739745
2016-03-13    1.599420
2016-03-14    3.662103
2016-03-15    3.361888
2016-03-16    2.906390
2016-03-17    3.048733
2016-03-18    1.283677
2016-03-19    3.511996
2016-03-20    3.814799
2016-03-21    3.731981
2016-03-22    3.253190
2016-03-23    3.219545
2016-03-24    2.901214
2016-03-25    3.051321
2016-03-26    3.310127
2016-03-27    3.139315
2016-03-28    3.535288
2016-03-29    3.400709
2016-03-30    3.304951
2016-03-31    3.141903
2016-04-01    3.460234
2016-04-02    3.628458
2016-04-03    3.913145
2016-04-04    3.690572
2016-04-05    1.335438
2016-04-06    0.326095
2016-04-07    0.150107
Name: date_crawled, dtype: float64


> In date_crawled column, each date evenly doing crawled from 2016-03-05 to 2016-04-07 and ranging from **0.1% to 3.7%**.

#### Get a percentage for ad_created column

In [17]:
ad_created = (
    autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False) * 100
).head(25)

print("Top 25 date ads created:")
print(ad_created)

Top 25 date ads created:
2016-04-03    3.944201
2016-03-20    3.825151
2016-03-21    3.760449
2016-04-04    3.731981
2016-03-12    3.713864
2016-04-02    3.594814
2016-03-28    3.537876
2016-03-14    3.496467
2016-04-01    3.444706
2016-03-19    3.405885
2016-03-29    3.400709
2016-03-07    3.380005
2016-03-15    3.343772
2016-03-26    3.320479
2016-03-11    3.304951
2016-03-10    3.297187
2016-03-30    3.289423
2016-03-09    3.266130
2016-03-08    3.260954
2016-03-22    3.232485
2016-03-23    3.193664
2016-03-31    3.157432
2016-03-27    3.126375
2016-03-25    3.066850
2016-03-17    3.017676
Name: ad_created, dtype: float64


> From this data, People sell their car heavily in **March 2016 to April 2016**.

#### Get a percentage for last_seen column

In [18]:
last_seen = (
    autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False) * 100
).sort_index()

print(last_seen)

2016-03-05     0.108698
2016-03-06     0.357152
2016-03-07     0.455498
2016-03-08     0.626310
2016-03-09     0.892880
2016-03-10     0.980874
2016-03-11     1.172391
2016-03-12     2.220554
2016-03-13     0.838531
2016-03-14     1.198271
2016-03-15     1.498486
2016-03-16     1.545071
2016-03-17     2.637232
2016-03-18     0.737597
2016-03-19     1.459665
2016-03-20     1.979865
2016-03-21     1.969513
2016-03-22     2.078211
2016-03-23     1.790937
2016-03-24     1.853050
2016-03-25     1.775408
2016-03-26     1.607184
2016-03-27     1.407904
2016-03-28     1.943632
2016-03-29     2.078211
2016-03-30     2.344781
2016-03-31     2.274904
2016-04-01     2.318901
2016-04-02     2.492301
2016-04-03     2.443127
2016-04-04     2.337017
2016-04-05    13.113693
2016-04-06    23.468516
2016-04-07    13.993633
Name: last_seen, dtype: float64


> **The ads heavily seen at 2016-04-05 to 2016-04-07**

#### Understanding the distribution registration year

This is an optional step, it is nice to know what year that most car is registered.

In [19]:
print(autos["registration_year"].describe())

count    38639.000000
mean      2005.676596
std         86.671288
min       1000.000000
25%       2001.000000
50%       2005.000000
75%       2009.000000
max       9999.000000
Name: registration_year, dtype: float64


In [20]:
print(autos["registration_year"].value_counts().sort_index())

1000    1
1001    1
1927    1
1929    1
1931    1
       ..
5911    1
6200    1
8888    1
9000    1
9999    2
Name: registration_year, Length: 91, dtype: int64


> We can see that in registration_year column has an invalid value, **There are years that exceeded 2019 and below 1927 (car not even invented that year)**. Next step will fix these registration year.

### 5. Dealing with Incorrect Registration Year Data

One thing that stands out from the exploration in the last screen is that the registration_year column contains some odd values:

- The minimum value is 1000, before cars were invented
- The maximum value is 9999, many years into the future

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2019 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

In [21]:
# Removing incorrect registration_year, using year between 1927 tp 2019
autos = autos[autos["registration_year"].between(1927, 2019)]

print("Top ten registration year:")
print((autos["registration_year"].value_counts(normalize=True) * 100).head(10))

Top ten registration year:
2005    7.210169
2006    6.865842
2004    6.751929
2003    6.412779
2007    5.848392
2008    5.706001
2002    5.527365
2009    5.377207
2001    5.348729
2000    5.180448
Name: registration_year, dtype: float64


> The cars that are going to be sold in ebay is registered mostly in **early 2000s**.

### 6. Exploring Price by Brand

When working with data on cars, it's natural to explore variations across different car brands. It can use aggregation to understand the brand column.

In [22]:
# Get the most popular brand
brand_percentage = autos["brand"].value_counts(normalize=True) * 100

print("Top 5 most car popular brand:")
print(brand_percentage.head())

Top 5 most car popular brand:
volkswagen       21.278414
bmw              12.377673
mercedes_benz    11.070264
audi              9.680008
opel              8.980997
Name: brand, dtype: float64


From that data, we got the 5 most popular cars in ebay: **VW, BMW, Mercedes Benz, Audi, Opel**.

Next we will agregate these brands with mean price in each brand.

In [23]:
# Aggregating brand with mean price
mean_price_brands = {}

top_five_brands = brand_percentage.head().index
for brand in top_five_brands:
    selected_rows = autos[autos["brand"] == brand]
    mean_price = selected_rows["price"].mean()
    mean_price_brands[brand] = mean_price

In [24]:
print(mean_price_brands)

{'volkswagen': 8149.5539603358075, 'bmw': 9311.7678309977, 'mercedes_benz': 9215.087932647333, 'audi': 10200.053490238031, 'opel': 7735.584318247334}


> From above analysis, we got mean price for top brand cars in ebay:
>1. **Volkswagen: \\$8,149**
>2. **BMW: \\$9,311**
>3. **Mercedes Benz: \\$9,215**
>4. **Audi: \\$10,200**
>5. **Opel: \\$7,735**

It is nice to know the relation between average mileage and mean price for each brand.

In [25]:
# Aggregating brand with mean price
avg_mileage_brands = {}

top_five_brands = brand_percentage.head().index
for brand in top_five_brands:
    selected_rows = autos[autos["brand"] == brand]
    age_mileage = selected_rows["odometer_km"].mean()
    avg_mileage_brands[brand] = age_mileage

In [38]:
# Creating pandas series for those columns
mean_price_series = pd.Series(mean_price_brands)
avg_mileage_series = pd.Series(avg_mileage_brands)

# Convert to pandas dataframe
mean_price_df = pd.DataFrame(mean_price_series, columns=["mean_price"])
avg_mileage_df = pd.DataFrame(avg_mileage_series, columns=["avg_mileage"])

# Concate the dataframe
price_vs_mileage = pd.concat([mean_price_df, avg_mileage_df], axis=1)
print(price_vs_mileage)

                 mean_price    avg_mileage
volkswagen      8149.553960  126064.606400
bmw             9311.767831  132121.941017
mercedes_benz   9215.087933  130164.873714
audi           10200.053490  127897.833645
opel            7735.584318  124132.314788


> #### It can be concluded, **that are no relation** between mean price vs average mileage for each car brand.